mixins/visibility-mixin.js

```diff
@@ -0,0 +1,207 @@
+
+const reduceMotion = matchMedia('(prefers-reduced-motion: reduce)').matches;
+const transitionDuration = 300;
+const moveYValue = 10;
+
+export const VisibilityMixin = dedupeMixin(superclass => class extends superclass {
+	static get properties() {
+		return {
+			animate: { type: String }
+		};
+	}
+
+	constructor() {
+		super();
+		this.dummy = document.createElement('div');
+		this.dummy.style.overflow = 'hidden';
+		this.dummy.style.display = 'grid';
+	}
+
+	firstUpdated() {
+		this.opacityOriginal = window.getComputedStyle(this).opacity;
+		this.transformOriginal = window.getComputedStyle(this).transform;
+		this.displayOriginal = window.getComputedStyle(this).display;
+		this.transitionOriginal = window.getComputedStyle(this).transition;
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+		if (changedProperties.has('animate')) {
+			if (this.animate === 'show') {
+				this._animateShow();
+			} else if (this.animate === 'hide') {
+				this._animateHide();
+			} else if (this.animate === 'remove') {
+				this._animateRemove();
+			}
+		}
+	}
+
+	_animateHide() {
+		const thisOnTransitionStart = () => {
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-hide-start',
+				{ bubbles: true, composed: false }
+			));
+		};
+		const dummyOnTransitionEnd = () => {
+			this.dummy.replaceWith(this);
+			this.style.display = 'none';
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-hide-end',
+				{ bubbles: true, composed: false }
+			));
+		};
+		this._animateHideRemove(thisOnTransitionStart, dummyOnTransitionEnd);
+	}
+
+	_animateHideRemove(thisOnTransitionStart, dummyOnTransitionEnd) {
+		const animateHideRemoveStyle = {
+			initial: {
+				transition: `all ${transitionDuration}ms ease`,
+				opacity: `${this.opacityOriginal}`,
+				transform: `${this.transformOriginal}`
+			},
+			initialDummy: {
+				transition: `height ${transitionDuration}ms ease ${transitionDuration / 3}ms`,
+				height: `calc(${this.scrollHeight}px + ${window.getComputedStyle(this).marginTop} + ${window.getComputedStyle(this).marginBottom})`
+			},
+			final: {
+				opacity: '0',
+				transform: `translateY(-${moveYValue}px)`
+			},
+			finalDummy: {
+				height: '0'
+			},
+			thisOnTransitionStart: thisOnTransitionStart,
+			dummyOnTransitionEnd: dummyOnTransitionEnd
+		};
+		this._animateVisibility(animateHideRemoveStyle);
+	}
+
+	_animateRemove() {
+		const thisOnTransitionStart = () => {
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-remove-start',
+				{ bubbles: true, composed: false }
+			));
+		};
+		const dummyOnTransitionEnd = () => {
+			this.dummy.replaceWith(this);
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-remove-end',
+				{ bubbles: true, composed: false }
+			));
+			this.remove();
+		};
+		this._animateHideRemove(thisOnTransitionStart, dummyOnTransitionEnd);
+	}
+
+	_animateShow() {
+		this.style.display = this.displayOriginal;
```

Comment:
I'm a little worried about all the code that's manipulating the DOM/style outside of the `render()` method. Technically in Lit we're not supposed to do this -- `render()` should be a function of the component's properties and all re-renders should occur by changing them. I'm not sure how else to accomplish this though...

Reply:
One option we could discuss around this would be for this to go back to being a container instead of a mixin -- I know we had briefly discussed that in Slack before this was written. That way it would have its own `render()` method so all these direct manipulations would go away.

The downside would be it would be wrapping `d2l-list-item`s, which might mess up the ARIA and other cases where list was expecting particular children. But maybe `role="presentation"` on itself and a few other tweaks could get around that though. Discuss!

Reply:
For list items, I think that it may be worth making a special type of `d2l-list-item` that combines the visibility container with the `d2l-list-item-mixin`. It might need a bit of adjustment to do this effectively, but this seems cleaner than wrapping list-item components themselves.

Reply:
**Alternatively**, if we stick with a mixin, I wonder if we could change to use render methods that take templates as arguments like what we did with our list item mixins: https://github.com/BrightspaceUI/core/blob/master/components/list/list-item-checkbox-mixin.js

They require the consumer to call the render methods, but this seems like a reasonable thing that we already do elsewhere. This gives us indirect access to the render method so that we can rely on properties instead of manipulating the DOM and styles manually.

Reply:
I like that, and Dave B suggested something similar. Let's continue this discussion the Slack thread in `#wc-visibility-animations`.